waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,512 @@
+  let (syncStart, syncEnd) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?serverInitialize(self.storage, conn, self.maxFrameSize, syncStart, syncEnd)
+
+  var sent = initialized
+
+  while true:
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = received.payload
+
+    let reconciled = (?received.serverReconcile()).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      let hashes = await completed.serverTerminate()
+
+      return ok(hashes)
+
+    sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = reconciled.payload
+
+proc initProtocolHandler(self: WakuSync) =
+  proc handle(conn: Connection, proto: string) {.async, closure.} =
+    debug "sync session requested",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+    let hashes = (await self.handleSyncSession(conn)).valueOr:
+      debug "sync session ended",
+        server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
+        error "transfer callback failed", error = $error
+
+    debug "sync session ended gracefully",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+  self.handler = handle
+  self.codec = WakuSyncCodec
+
+proc createPruneCallback(
+    self: WakuSync, wakuArchive: WakuArchive
+): Result[PruneCallBack, string] =
+  if wakuArchive.isNil():
+    return err ("waku archive unavailable")
+
+  let callback: PruneCallback = proc(
+      pruneStart: Timestamp, pruneStop: Timestamp, cursor: Option[WakuMessageHash]
+  ): Future[
+      Result[(seq[(WakuMessageHash, Timestamp)], Option[WakuMessageHash]), string]
+  ] {.async: (raises: []), closure.} =
+    let archiveCursor =
+      if cursor.isSome():
+        some(cursor.get())
+      else:
+        none(ArchiveCursor)
+
+    let query = ArchiveQuery(
+      includeData: true,
+      cursor: archiveCursor,
+      startTime: some(pruneStart),
+      endTime: some(pruneStop),
+      pageSize: 100,
+    )
+
+    let catchable = catch:
+      await wakuArchive.findMessages(query)
+
+    if catchable.isErr():
+      return err("archive error: " & catchable.error.msg)
+
+    let res = catchable.get()
+    let response = res.valueOr:
+      return err("archive error: " & $error)
+
+    let elements = collect(newSeq):
+      for (hash, msg) in response.hashes.zip(response.messages):
+        (hash, msg.timestamp)
+
+    let cursor = response.cursor
+
+    return ok((elements, cursor))
+
+  return ok(callback)
+
+proc createTransferCallback(
```

Comment:
I think we don't need neither this proc or `createPruneCallback`. Having the local private procs will suffice

Reply:
see my reasoning above